File: sdk/ts/src/output/flags.ts
/**
 * @module output/flags
 *
 * Commander integration: registerOutputFlags adds the standard output
 * flag suite (--format, --format-opt, --format-help, --cols, --template,
 * --output|-o) to a Commander program. Mirrors Go's RegisterFlagsWith.
 */

import type { Command } from 'commander';
import { Registry, defaultRegistry } from './registry';

/** Internal stash mapping a Commander program → Registry. */
const registryStash = new WeakMap<Command, Registry>();

/** Returns the Registry attached to cmd (walks parents) or defaultRegistry. */
export function registryFor(cmd: Command): Registry {
  for (let c: Command | null = cmd; c; c = (c as { parent?: Command | null }).parent ?? null) {
    const r = registryStash.get(c);
    if (r) return r;
  }
  return defaultRegistry;
}

/** Options for registerOutputFlags. */
export interface RegisterOutputFlagsOptions {
  /** Custom registry (default: defaultRegistry). */
  registry?: Registry;
  /** Per-flag opt-out. */
  disable?: {
    format?: boolean;
    formatOpt?: boolean;
    formatHelp?: boolean;
    cols?: boolean;
    template?: boolean;
    output?: boolean;
  };
}

/**
 * Adds the output flag suite to a Commander program. Each flag is opt-out
 * via `opts.disable.<flag>`. The active registry can be swapped via
 * `opts.registry`; tests + multi-CLI binaries use this for isolated sets.
 */
export function registerOutputFlags(
  program: Command,
  opts: RegisterOutputFlagsOptions = {},
): void {
  const registry = opts.registry ?? defaultRegistry;
  const d = opts.disable ?? {};

  if (registry !== defaultRegistry) {
    registryStash.set(program, registry);
  }

  if (!d.format) {
    program.option(
      '--format <fmt>',
      `Output format (${registry.keys().join(', ')})`,
      'table',
    );
  }
  // --format is the only output flag in the cross-language parity FLAGS
  // contract; everything below is kit plumbing hidden from default --help.
  // Adopters opting in via --help-all still see them (Commander's
  // showHidden flag is wired by applyHelpTheme).
  if (!d.formatOpt) {
    const opt = program
      .createOption(
        '--format-opt <kv...>',
        'Per-format option as key=value (repeatable; bool keys may omit =value)',
      )
      .default([] as string[])
      .argParser(collectKv);
    opt.hidden = true;
    program.addOption(opt);
  }
  if (!d.formatHelp) {
    const opt = program.createOption(
      '--format-help [fmt]',
      'Show available formats and their options (use --format-help <key> for one)',
    );
    opt.hidden = true;
    program.addOption(opt);
  }
  if (!d.cols) {
    const colsOpt = program
      .createOption('--cols <cols...>', 'Restrict columns to this comma-separated list (repeatable)')
      .default([] as string[])
      .argParser(collectCols);
    colsOpt.hidden = true;
    program.addOption(colsOpt);
    const columnsOpt = program
      .createOption('--columns <cols...>', 'Alias for --cols')
      .default([] as string[])
      .argParser(collectCols);
    columnsOpt.hidden = true;
    program.addOption(columnsOpt);
  }
  if (!d.template) {
    const opt = program.createOption(
      '--template <tpl>',
      'eta template applied to results (mutually exclusive with --cols)',
    );
    opt.hidden = true;
    program.addOption(opt);
  }
  if (!d.output) {
    const opt = program
      .createOption('-o, --output <path>', 'Write output to path (use - or empty for stdout)')
      .default('');
    opt.hidden = true;
    program.addOption(opt);
  }
}

/** Variadic collector for --format-opt. Each invocation appends one value. */
function collectKv(value: string, prev: readonly string[]): string[] {
  return [...prev, value];
}

/**
 * Variadic collector for --cols / --columns. Splits each invocation on
 * commas so callers can pass either repeated flags or a single
 * comma-separated string. Dedupe happens later in dispatch.
 */
function collectCols(value: string, prev: readonly string[]): string[] {
  return [...prev, ...value.split(',')];
}

/**
 * Merges --cols + --columns from a Commander opts object into a single
 * ordered, deduped string[]. Trims whitespace, drops empty entries.
 */
export function resolveCols(opts: Record<string, unknown>): string[] {
  const raw: string[] = [];
  for (const k of ['cols', 'columns']) {
    const v = opts[k];
    if (Array.isArray(v)) raw.push(...(v as string[]));
  }
  const seen = new Set<string>();
  const out: string[] = [];
  for (const item of raw) {
    for (const part of item.split(',')) {
      const p = part.trim();
      if (p === '' || seen.has(p)) continue;
      seen.add(p);
      out.push(p);
    }
  }
  return out;
}

File: go/console/output/flags.go
package output

import (
	"strings"
	"sync"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

// registryStash maps *cobra.Command → *Registry for commands wired with
// WithRegistry. Dispatch uses this to recover the active registry when it
// differs from Default.
var registryStash sync.Map

// Viper keys bound by RegisterFlagsWith. Callers that read these keys
// after flag parsing should use these constants to avoid typos.
const (
	flagFormat     = "format"
	flagFormatOpt  = "format-opt"
	flagFormatHelp = "format-help"
	flagCols       = "cols"
	flagColumns    = "columns"
	flagTemplate   = "template"
	flagOutput     = "output"
)

// RegistryOption customizes RegisterFlagsWith. Use the With* / Disable*
// helpers below; do not construct registryOptions directly.
type RegistryOption func(*registryOptions)

type registryOptions struct {
	registry      *Registry
	disableOutput bool
}

// WithRegistry binds Dispatch / format-help to a specific Registry rather
// than the package-level Default. Useful for tests and multi-CLI binaries
// that want isolated formatter sets.
func WithRegistry(r *Registry) RegistryOption {
	return func(o *registryOptions) { o.registry = r }
}

// DisableOutputFlag suppresses registration of the --output / -o flag. Use
// for commands or sub-CLIs that must always write to stdout (e.g. when
// stdout is part of a pipeline contract).
func DisableOutputFlag() RegistryOption {
	return func(o *registryOptions) { o.disableOutput = true }
}

// RegisterFlagsWith adds the output persistent flags to cmd and binds
// them to viper keys. Pass options to swap registries or suppress
// individual flags.
//
// Flags registered (all persistent on cmd):
//   - --format <key>           default "table"; bound to "format"
//   - --format-opt key=value   StringSlice, repeatable; bound to "format-opt"
//   - --format-help [format]   bool toggle; bound to "format-help"
//   - --cols, --columns        StringSlice, repeatable + comma-split;
//     bound to "cols" (and "columns" alias)
//   - --template <go-tmpl>     string; bound to "template"
//   - --output, -o <path>      string; bound to "output"; "" or "-" = stdout
//
// The registry used for format lookup defaults to Default; pass
// WithRegistry to override. Pass DisableOutputFlag to suppress -o.
func RegisterFlagsWith(cmd *cobra.Command, v *viper.Viper, opts ...RegistryOption) {
	o := &registryOptions{registry: Default}
	for _, apply := range opts {
		apply(o)
	}

	pf := cmd.PersistentFlags()

	pf.String(flagFormat, Table, "Output format ("+strings.Join(o.registry.Keys(), ", ")+")")
	_ = v.BindPFlag(flagFormat, pf.Lookup(flagFormat))

	pf.StringSlice(flagFormatOpt, nil,
		"Per-format option as key=value (repeatable; bool keys may omit =value)")
	_ = v.BindPFlag(flagFormatOpt, pf.Lookup(flagFormatOpt))

	pf.Bool(flagFormatHelp, false,
		"Show available formats and their options (use --format <key> --format-help for one)")
	_ = v.BindPFlag(flagFormatHelp, pf.Lookup(flagFormatHelp))

	pf.StringSlice(flagCols, nil,
		"Restrict columns to this comma-separated list (repeatable)")
	_ = v.BindPFlag(flagCols, pf.Lookup(flagCols))

	// --columns is a strict alias of --cols. Both bind to the same viper
	// key so callers can read either.
	pf.StringSlice(flagColumns, nil, "Alias for --cols")
	_ = v.BindPFlag(flagColumns, pf.Lookup(flagColumns))

	pf.String(flagTemplate, "",
		"Go text/template applied to results (mutually exclusive with --cols)")
	_ = v.BindPFlag(flagTemplate, pf.Lookup(flagTemplate))

	if !o.disableOutput {
		pf.StringP(flagOutput, "o", "",
			"Write output to path (use - or empty for stdout)")
		_ = v.BindPFlag(flagOutput, pf.Lookup(flagOutput))
	}

	// Stash the resolved registry so Dispatch can recover it without
	// threading RegistryOption through every callsite.
	if o.registry != Default {
		registryStash.Store(cmd, o.registry)
	}
}

// registryFor walks cmd's parent chain and returns the registry stashed
// on the nearest ancestor (or Default if none).
func registryFor(cmd *cobra.Command) *Registry {
	for c := cmd; c != nil; c = c.Parent() {
		if v, ok := registryStash.Load(c); ok {
			if r, ok := v.(*Registry); ok {
				return r
			}
		}
	}
	return Default
}

// resolveCols merges --cols + --columns values into a single ordered,
// deduped slice. Each value may itself be comma-separated (cobra
// StringSlice does not split on comma when values arrive via viper
// config). When cmd is non-nil and the flag was changed on the command
// line, that takes precedence over viper — keeping behavior consistent
// with adopters that pass an unbound viper to Dispatch.
func resolveCols(cmd *cobra.Command, v *viper.Viper) []string {
	raw := append([]string{}, lookupStringSlice(cmd, v, flagCols)...)
	raw = append(raw, lookupStringSlice(cmd, v, flagColumns)...)
	seen := make(map[string]struct{}, len(raw))
	out := make([]string, 0, len(raw))
	for _, item := range raw {
		for _, part := range strings.Split(item, ",") {
			p := strings.TrimSpace(part)
			if p == "" {
				continue
			}
			if _, dup := seen[p]; dup {
				continue
			}
			seen[p] = struct{}{}
			out = append(out, p)
		}
	}
	return out
}
